File: oops.py
from abc import ABC, abstractmethod

# 1. CLASS & OBJECT
# A class is a blueprint. An object is an instance of the class.
class Car:
    def __init__(self, brand, color):
        self.brand = brand
        self.color = color

    def drive(self):
        print(f"The {self.color} {self.brand} is driving.")

# 2. ENCAPSULATION
# Encapsulation hides internal data using private variables and provides access through methods.
class BankAccount:
    def __init__(self):
        self.__balance = 0  # private variable (double underscore)

    def deposit(self, amount):
        self.__balance += amount  # internal logic is hidden

    def get_balance(self):
        return self.__balance  # access controlled via method

# 3. INHERITANCE
# A child class inherits behavior from the parent class and can override it.
class Animal:
    def speak(self):
        print("Animal makes a sound")

class Dog(Animal):
    def speak(self):
        print("Dog barks")  # overriding parent method

# 4. POLYMORPHISM
# Polymorphism allows the same method name to behave differently based on the object.
class Cat:
    def sound(self):
        print("Meow")

class AnotherDog:
    def sound(self):
        print("Bark")

def make_sound(animal):
    animal.sound()  # Works for any object with a sound() method

# Example usage of polymorphism
cat = Cat()
dog = AnotherDog()

ex=make_sound(cat).sound()  # Output: Meow 
ex2=make_sound(dog).sound()  # Output: Bark


# 5. ABSTRACTION
# Abstraction hides complex implementation and shows only essential features.
class Shape(ABC):  # Abstract Base Class
    @abstractmethod
    def area(self):
        pass  # no implementation here

class Circle(Shape):
    def area(self):
        radius = 5
        return 3.14 * radius * radius  # implementation provided by subclass

# MAIN PROGRAM
if __name__ == "__main__":
    print("=== CLASS & OBJECT ===")
    my_car = Car("Toyota", "Red")
    my_car.drive()  # Output: The Red Toyota is driving.

    print("\n=== ENCAPSULATION ===")
    account = BankAccount()
    account.deposit(100)
    print("Balance:", account.get_balance())  # Output: Balance: 100

    print("\n=== INHERITANCE ===")
    a = Animal()
    d = Dog()
    a.speak()  # Output: Animal makes a sound
    d.speak()  # Output: Dog barks

    print("\n=== POLYMORPHISM ===")
    make_sound(Cat())       # Output: Meow
    make_sound(AnotherDog())  # Output: Bark

    print("\n=== ABSTRACTION ===")
    c = Circle()
    print("Area of circle:", c.area())  # Output: 78.5

File: oops.java
// 1. CLASS & OBJECT
// A class is a blueprint; objects are created from that blueprint.
class Car {
    String brand;
    String color;

    Car(String brand, String color) {
        this.brand = brand;
        this.color = color;
    }

    void drive() {
        System.out.println("The " + color + " " + brand + " is driving.");
    }
}

// 2. ENCAPSULATION
// Encapsulation protects data using private fields and public getters/setters.
class BankAccount {
    private int balance = 0;

    void deposit(int amount) {
        balance += amount;
    }

    int getBalance() {
        return balance;
    }
}

// 3. INHERITANCE
// Inheritance allows one class to acquire properties and methods of another.
class Animal {
    void speak() {
        System.out.println("Animal makes a sound");
    }
}

class Dog extends Animal {
    @Override
    void speak() {
        System.out.println("Dog barks");  // Method overriding
    }
}

// 4. POLYMORPHISM
// Polymorphism allows objects to take many forms; interfaces help achieve this.
interface Soundable {
    void sound();
}

class Cat implements Soundable {
    public void sound() {
        System.out.println("Meow");
    }
}

class AnotherDog implements Soundable {
    public void sound() {
        System.out.println("Bark");
    }
}

class SoundTest {
    static void makeSound(Soundable animal) {
        animal.sound();  // Calls sound() on any Soundable object
    }
}

// 5. ABSTRACTION
// Abstract classes define method signatures without implementations.
abstract class Shape {
    abstract double area();  // Abstract method
}

class Circle extends Shape {
    double area() {
        double radius = 5;
        return 3.14 * radius * radius;  // Concrete implementation
    }
}

// MAIN CLASS TO RUN EVERYTHING
public class OOPConcepts {
    public static void main(String[] args) {
        System.out.println("=== CLASS & OBJECT ===");
        Car car = new Car("Toyota", "Red");
        car.drive();

        System.out.println("\n=== ENCAPSULATION ===");
        BankAccount account = new BankAccount();
        account.deposit(100);
        System.out.println("Balance: " + account.getBalance());

        System.out.println("\n=== INHERITANCE ===");
        Animal a = new Animal();
        Dog d = new Dog();
        a.speak();  // Animal makes a sound
        d.speak();  // Dog barks

        System.out.println("\n=== POLYMORPHISM ===");
        SoundTest.makeSound(new Cat());         // Meow
        SoundTest.makeSound(new AnotherDog());  // Bark

        System.out.println("\n=== ABSTRACTION ===");
        Shape c = new Circle();
        System.out.println("Area of circle: " + c.area());  // 78.5
    }
}
